scalafix-core/shared/src/main/scala/scalafix/internal/patch/EscapeHatch.scala

```diff
@@ -33,49 +33,49 @@ class EscapeHatch private (
     val usedEscapes = mutable.Set.empty[EscapeFilter]
     val lintMessages = List.newBuilder[LintMessage]
 
-    def isDisabledByEscape(name: RuleName, start: Int): Boolean =
-      // annotatedEscapes takes precedence over anchoredEscapes
-      annotatedEscapes.isEnabled(name, start) match {
-        case (false, Some(culprit)) => // if disabled, there must be a escape
-          usedEscapes += culprit
-          true
-        case _ =>
-          val (enabled, culprit) = anchoredEscapes.isEnabled(name, start)
-          culprit.foreach(escape => usedEscapes += escape)
-          !enabled
-      }
+    def isDisabled(name: RuleName, pos: Position): Boolean = {
+      def isDisabledByEscape() =
+        // annotatedEscapes takes precedence over anchoredEscapes
+        annotatedEscapes.isEnabled(name, pos.start) match {
+          case (false, Some(culprit)) => // if disabled, there must be a escape
+            usedEscapes += culprit
+            true
+          case _ =>
+            val (enabled, culprit) = anchoredEscapes.isEnabled(name, pos.start)
+            culprit.foreach(escape => usedEscapes += escape)
+            !enabled
+        }
+
+      val byGit = diff.isDisabled(pos)
+      val byEscape = isDisabledByEscape()
+      byGit || byEscape
+    }
 
     def loop(name: RuleName, patch: Patch): Patch = patch match {
       case AtomicPatch(underlying) =>
         val hasDisabledPatch = {
           val patches = Patch.treePatchApply(underlying)(ctx, index)
-          patches.exists { tp =>
-            val byGit = diff.isDisabled(tp.tok.pos)
-            val byEscape = isDisabledByEscape(name, tp.tok.pos.start)
-            byGit || byEscape
-          }
+          patches.exists(tp => isDisabled(name, tp.tok.pos))
         }
 
-        if (hasDisabledPatch) EmptyPatch
-        else loop(name, underlying)
+        if (hasDisabledPatch) EmptyPatch else loop(name, underlying)
 
       case Concat(a, b) =>
         Concat(loop(name, a), loop(name, b))
 
       case LintPatch(orphanLint) =>
         val lint = orphanLint.withOwner(name)
-
-        val byGit = diff.isDisabled(lint.position)
-        val byEscape = isDisabledByEscape(lint.id, lint.position.start)
-
-        val isLintDisabled = byGit || byEscape
+        val isLintDisabled = isDisabled(lint.id, lint.position)
 
         if (!isLintDisabled) {
           lintMessages += lint
         }
 
         EmptyPatch
 
+      case tp: TokenPatch =>
+        if (isDisabled(name, tp.tok.pos)) EmptyPatch else tp
```

Comment:
> fail testkit tests if a rule does emits top-level non-AtomicPatch

Reply:
Whoops, that was not supposed to go out right away.

Reply:
Shift+enter is submitting my comments 😅  will respond in a bit

Reply:
> Do you mean 2 dependent patches which are atomic? Or not atomic?

Not atomic. Atomic is a recent feature that has not been mandated so far, which means that existing rules do not implement it.

> Rules can (and do) return multiple independent patches from fix and users should be able to suppress them individually.

That assumes the rule author knows about `.atomic`, those individual patches could be dependent so I would be careful to implicitly assume `.atomic` for each of them.

> If you prefer to discuss this in a separate issue/PR, I can revert to the initial solution for issue 1 so that we get the issue addressed forMissingFinal.

I think a separate issue for this particular fix would be best.

> However, I'd argue that the rule has a bug which prevents it from being correctly suppressed.

I agree, that is why I propose we fail unit tests if a rule does not implement `.atomic`. This forces rule authors to document which parts are `.atomic` while developing a rule.

Reply:
Ok, got you. I will remove the commit that changed `EscapeHatch`

Reply:
done